pkg/obfuscator/domain.go

```diff
@@ -2,102 +2,77 @@ package obfuscator
+	return d.findDomains(s)
+}
 
+func (d *domainObfuscator) findDomains(input string) string {
 	output := input
-	for _, domain := range domains {
-		// if this is an IP address then do nothing
-		if d.ipPattern.MatchString(domain) {
-			continue
+	for _, p := range d.domainPatterns {
+		matches := p.FindAllStringSubmatch(output, -1)
+		for _, m := range matches {
+			if len(m) != 3 {
+				continue
+			}
+			baseDomain := m[2]
+			subDomain := m[1]
+			obfuscatedBaseDomain := d.obfuscatedDomain(baseDomain)
+			var replacement string
+			if subDomain != "" {
+				replacement = fmt.Sprintf("%s%s", subDomain, obfuscatedBaseDomain)
+			} else {
+				replacement = obfuscatedBaseDomain
+			}
+			output = strings.ReplaceAll(output, m[0], replacement)
+			d.ReportReplacement(m[0], replacement)
 		}
-		output = strings.ReplaceAll(output, domain, d.obfuscatedDomain(domain))
 	}
 	return output
 }
 
-func (d *domainObfuscator) Contents(s string) string {
-	return d.findDomains(s)
-}
-
 func (d *domainObfuscator) obfuscatedDomain(domain string) string {
-	parts := strings.Split(domain, ".")
-	var (
-		mainDomain   string
-		subDomain    string
-		hasExtension bool
-	)
-	// if the last part is present in known list of tlds then combine last and second-to-last parts to get
-	// the main domain otherwise the last part is the main domain
-	if _, ok := d.tlds[parts[len(parts)-1]]; ok {
-		mainDomain = strings.Join(parts[len(parts)-2:], ".")
-		subDomain = strings.Join(parts[:len(parts)-2], ".")
-		hasExtension = true
-	} else {
-		mainDomain = parts[len(parts)-1]
-		subDomain = strings.Join(parts[:len(parts)-1], ".")
-	}
-
-	var (
-		obfuscatedMainDomain string
-		ok                   bool
-	)
-	// if the obfuscated main domain is present then use it, otherwise generate one and store it.
-	if obfuscatedMainDomain, ok = d.mainDomains[mainDomain]; !ok {
-		d.domainCount++
-		if hasExtension {
-			obfuscatedMainDomain = fmt.Sprintf("obfuscated%04d.ext", d.domainCount)
-		} else {
-			obfuscatedMainDomain = fmt.Sprintf("obfuscated%04d", d.domainCount)
-		}
-		d.mainDomains[mainDomain] = obfuscatedMainDomain
+	if replacement, ok := d.domainMapping[domain]; ok {
+		return replacement
 	}
-
-	var obfuscatedDomain string
-	// if there is no subdomain then leave it out
-	if subDomain != "" {
-		obfuscatedDomain = fmt.Sprintf("%s.%s", subDomain, obfuscatedMainDomain)
-	} else {
-		obfuscatedDomain = obfuscatedMainDomain
-	}
-	d.ReportReplacement(domain, obfuscatedDomain)
-	return obfuscatedDomain
+	d.domainCount++
+	replacement := fmt.Sprintf(obfuscatedTemplate, d.domainCount)
+	d.domainMapping[domain] = replacement
+	return replacement
 }
 
-func NewDomainObfuscator(tlds []string) Obfuscator {
-	tldSet := map[string]struct{}{}
-	for _, tld := range tlds {
-		tldSet[strings.TrimLeft(tld, ".")] = struct{}{}
+func NewDomainObfuscator(domains []string) (Obfuscator, error) {
+	patterns := make([]*regexp.Regexp, len(domains))
```

Comment:
now that we have the static and consistent types, can you also add that behaviour in here?

Reply:
In a separate PR. Right now this is only consistent replacement.